src/choker.cpp

```diff
@@ -168,13 +158,24 @@ namespace {
 		//   |             V             |
 		//   +---------------------------+
 		//   0%    num have pieces     100%
-		int const t1_total = t1->torrent_file().num_pieces();
-		int const t2_total = t2->torrent_file().num_pieces();
-		int const score1 = (lhs->num_have_pieces() < t1_total / 2
-			? t1_total - lhs->num_have_pieces() : lhs->num_have_pieces()) * 1000 / t1_total;
-		int const score2 = (rhs->num_have_pieces() < t2_total / 2
-			? t2_total - rhs->num_have_pieces() : rhs->num_have_pieces()) * 1000 / t2_total;
+		std::shared_ptr<torrent> const t = peer->associated_torrent().lock();
+		TORRENT_ASSERT(t);
+
+		std::int64_t const total_size = t->torrent_file().total_size();
+		std::int64_t have_size = peer->num_have_pieces() * t->torrent_file().piece_length();
+		have_size = std::max(have_size, peer->uploaded_at_last_round());
```

Comment:
why is `uploaded_at_last_round()` better than `statistics().total_payload_upload()`?

Reply:
`statistics().total_payload_upload()` is assigned to `m_uploaded_at_last_round` in `peer_connection::reset_choke_counters()` before unchoking, so they should be exactly the same.